pkg/subsystem/linux/names.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2023 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package linux
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/google/syzkaller/pkg/subsystem/entity"
+)
+
+// setSubsystemNames assigns unique names to the presented subsystems.
+// If it failed to assign a name to a subsystem, the Name field remains empty.
+func setSubsystemNames(list []*entity.Subsystem) error {
+	dups := map[string]bool{}
+	for _, item := range list {
+		if item.Name == "" {
+			continue
+		}
+		if dups[item.Name] {
+			return fmt.Errorf("duplicate subsystem name: %s", item.Name)
+		}
+		dups[item.Name] = true
+	}
+	for _, item := range list {
+		if item.Name != "" {
+			continue
+		}
+		// For now, we can only infer name from the list email.
+		if len(item.Lists) == 0 {
+			return fmt.Errorf("no emailinfo for %#v", item)
+		}
+		email := item.Lists[0]
+		name := emailToName(email)
+		if name == "" || !validateName(name) {
+			return fmt.Errorf("failed to extract a name from %s", email)
+		}
+		if !dups[name] {
+			item.Name = name
+			dups[name] = true
+			continue
+		}
+		// As a fallback, assign the full email address as the name, with the hope that
+		// the reviewer would notice and take action.
```

Comment:
Yes, after all the previous refactorings of the PR we no longer assign names and always. It will come back in one of the future PRs though (when we add the custom rules for MAINTAINERS records without lists).

For empty Lists, I still think we'd better add a check and return an error than let it panic.

Otherwise, I've simplified the code as you said.

Reply:
> It will come back in one of the future PRs though (when we add the custom rules for MAINTAINERS records without lists).

There may be other ways to fit that into the code. For example, instead of splitting assignment of names to 2 separate passes, we could do all name assignment inside of groupByList. The main piece of logic provided by this file now is emailToName and we will reuse it, but setSubsystemNames is not doing much now besides doing another loop over the list of subsystems, so it's logic may be merged into another loop over all subsystems.